File: nexus/src/main/java/io/xj/nexus/mixer/ActiveAudio.java
// Copyright (c) XJ Music Inc. (https://xjmusic.com) All Rights Reserved.

package io.xj.nexus.mixer;

import io.xj.hub.InstrumentConfig;
import io.xj.hub.tables.pojos.Instrument;
import io.xj.hub.tables.pojos.InstrumentAudio;
import io.xj.nexus.model.SegmentChoiceArrangementPick;
import jakarta.annotation.Nullable;

import java.util.Optional;
import java.util.UUID;

public class ActiveAudio {
  private final InstrumentConfig instrumentConfig;
  private final SegmentChoiceArrangementPick pick;
  private final InstrumentAudio audio;
  private final Long startAtMixerMicros;
  @Nullable
  private final Long stopAtMixerMicros;
  private final Instrument instrument;
  private final float amplitude;

  public ActiveAudio(
    SegmentChoiceArrangementPick pick,
    Instrument instrument,
    InstrumentAudio audio,
    Long startAtMixerMicros,
    @Nullable Long stopAtMixerMicros,
    float intensityAmplitude
  ) {
    this.pick = pick;
    this.audio = audio;
    this.startAtMixerMicros = startAtMixerMicros;
    this.stopAtMixerMicros = stopAtMixerMicros;
    this.instrument = instrument;

    // computed
    this.amplitude = intensityAmplitude * pick.getAmplitude() * instrument.getVolume() * audio.getVolume();
    this.instrumentConfig = new InstrumentConfig(instrument);
  }

  public UUID getId() {
    return pick.getId();
  }

  public SegmentChoiceArrangementPick getPick() {
    return pick;
  }

  public Instrument getInstrument() {
    return instrument;
  }

  public Long getStartAtMixerMicros() {
    return startAtMixerMicros;
  }

  public Optional<Long> getStopAtMixerMicros() {
    return Optional.ofNullable(stopAtMixerMicros);
  }

  public InstrumentAudio getAudio() {
    return audio;
  }

  public int getReleaseMillis() {
    return instrumentConfig.getReleaseMillis();
  }

  public float getAmplitude() {
    return amplitude;
  }
}

File: engine/src/audio/ActiveAudio.cpp
// Copyright (c) XJ Music Inc. (https://xjmusic.com) All Rights Reserved.

#include "xjmusic/audio/ActiveAudio.h"

using namespace XJ;

ActiveAudio::ActiveAudio(
    const SegmentChoiceArrangementPick *pick,
    const Instrument *instrument,
    const InstrumentAudio *audio,
    const unsigned long long startAtChainMicros,
    const unsigned long long stopAtChainMicros,
    const float fromIntensity,
    const float toIntensity) {
  this->pick = pick;
  this->audio = audio;
  this->startAtChainMicros = startAtChainMicros;
  this->stopAtChainMicros = stopAtChainMicros;
  this->instrument = instrument;

  // computed
  this->fromVolume = fromIntensity * pick->amplitude * instrument->volume * audio->volume;
  this->toVolume = toIntensity * pick->amplitude * instrument->volume * audio->volume;
}

UUID ActiveAudio::getId() const {
  return pick->id;
}

const SegmentChoiceArrangementPick *ActiveAudio::getPick() const {
  return pick;
}

const Instrument *ActiveAudio::getInstrument() const {
  return instrument;
}

unsigned long long ActiveAudio::getStartAtChainMicros() const {
  return startAtChainMicros;
}

unsigned long long ActiveAudio::getStopAtChainMicros() const {
  return stopAtChainMicros;
}

const InstrumentAudio *ActiveAudio::getAudio() const {
  return audio;
}

int ActiveAudio::getReleaseMillis() const {
  return instrument->config.releaseMillis;
}

float ActiveAudio::getAmplitude(const float ap) const {
  if (fromVolume == toVolume) return fromVolume;
  return fromVolume + (toVolume - fromVolume) * ap;
}

float ActiveAudio::getFromVolume() const {
  return fromVolume;
}

float ActiveAudio::getToVolume() const {
  return toVolume;
}
